File: integrasjonspunkt/src/main/java/no/difi/meldingsutveksling/nextmove/NextMoveDpiRequest.java
package no.difi.meldingsutveksling.nextmove;

import lombok.RequiredArgsConstructor;
import no.difi.begrep.sdp.schema_v10.SDPSikkerhetsnivaa;
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties;
import no.difi.meldingsutveksling.dpi.Document;
import no.difi.meldingsutveksling.dpi.MeldingsformidlerRequest;
import no.difi.meldingsutveksling.nextmove.message.OptionalCryptoMessagePersister;
import no.difi.meldingsutveksling.serviceregistry.externalmodel.ServiceRecord;
import no.difi.sdp.client2.domain.MetadataDokument;
import no.difi.sdp.client2.domain.digital_post.Sikkerhetsnivaa;
import no.difi.sdp.client2.domain.fysisk_post.Posttype;
import no.difi.sdp.client2.domain.fysisk_post.Returhaandtering;
import no.difi.sdp.client2.domain.fysisk_post.Utskriftsfarge;
import org.springframework.util.StringUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.Clock;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public class NextMoveDpiRequest implements MeldingsformidlerRequest {

    private static final String MISSING_TXT = "Missing title";

    private final IntegrasjonspunktProperties props;
    private final Clock clock;
    private final NextMoveMessage message;
    private final ServiceRecord serviceRecord;
    private final OptionalCryptoMessagePersister optionalCryptoMessagePersister;

    @Override
    public Document getDocument() {
        BusinessMessageFile primary = message.getFiles().stream()
                .filter(BusinessMessageFile::getPrimaryDocument)
                .findFirst()
                .orElseThrow(() -> new NextMoveRuntimeException("No primary documents found, aborting send"));

        String title = StringUtils.hasText(primary.getTitle()) ? primary.getTitle() : MISSING_TXT;
        return new Document(getContent(primary.getIdentifier()), primary.getMimetype(), primary.getFilename(), title);
    }

    @Override
    public List<Document> getAttachments() {
        return message.getFiles()
                .stream()
                .filter(f -> !f.getPrimaryDocument())
                .filter(f -> !isMetadataFile(f.getFilename()))
                .map(this::createDocument)
                .collect(Collectors.toList());
    }

    private Document createDocument(BusinessMessageFile file) {
        String title = StringUtils.hasText(file.getTitle()) ? file.getTitle() : MISSING_TXT;
        Document document = new Document(getContent(file.getIdentifier()), file.getMimetype(), file.getFilename(), title);
        if (isDigitalMessage() && getDigitalMessage().getMetadataFiler().containsKey(file.getFilename())) {
            String metadataFilename = getDigitalMessage().getMetadataFiler().get(file.getFilename());
            BusinessMessageFile metadataFile = message.getFiles().stream()
                    .filter(f -> f.getFilename().equals(metadataFilename))
                    .findFirst()
                    .orElseThrow(() -> new NextMoveRuntimeException(
                            String.format("Metadata document \"%s\" specified for \"%s\", but is not attached", metadataFilename, file.getFilename())));
            MetadataDokument md = new MetadataDokument(metadataFilename, metadataFile.getMimetype(), getContent(metadataFile.getIdentifier()));
            document.setMetadataDokument(md);
        }
        return document;
    }

    private byte[] getContent(String fileName) {
        try {
            return optionalCryptoMessagePersister.read(message.getMessageId(), fileName);
        } catch (IOException e) {
            throw new NextMoveRuntimeException(String.format("Could not read file \"%s\"", fileName), e);
        }
    }

    private boolean isMetadataFile(String filename) {
        if (isPrintMessage()) {
            return false;
        }
        return getDigitalMessage().getMetadataFiler().containsValue(filename);
    }

    private boolean isDigitalMessage() {
        return this.message.getBusinessMessage() instanceof DpiDigitalMessage;
    }

    private boolean isPrintMessage() {
        return this.message.getBusinessMessage() instanceof DpiPrintMessage;
    }

    private DpiDigitalMessage getDigitalMessage() {
        return (DpiDigitalMessage) message.getBusinessMessage();
    }

    private DpiPrintMessage getPrintMessage() {
        return (DpiPrintMessage) message.getBusinessMessage();
    }

    @Override
    public String getMottakerPid() {
        return message.getReceiverIdentifier();
    }

    @Override
    public String getSubject() {
        if (message.getBusinessMessage() instanceof DpiDigitalMessage) {
            return ((DpiDigitalMessage) message.getBusinessMessage()).getTittel();
        }
        if (message.getBusinessMessage() instanceof DpiPrintMessage) {
            return null;
        }
        throw new NextMoveRuntimeException(String.format("BusinessMessage not instance of either %s or %s", DpiDigitalMessage.class.getName(), DpiPrintMessage.class.getName()));
    }

    @Override
    public String getSenderOrgnumber() {
        return message.getSenderIdentifier();
    }

    @Override
    public Optional<String> getOnBehalfOfOrgnr() {
        return message.getSbd().getOnBehalfOfOrgNr();
    }

    @Override
    public Optional<String> getAvsenderIdentifikator() {
        if (isDigitalMessage()) {
            return Optional.ofNullable(getDigitalMessage().getAvsenderId());
        }
        return Optional.empty();
    }

    @Override
    public String getConversationId() {
        return message.getMessageId();
    }

    @Override
    public String getPostkasseAdresse() {
        return serviceRecord.getPostkasseAdresse();
    }

    @Override
    public byte[] getCertificate() {
        return serviceRecord.getPemCertificate().getBytes(StandardCharsets.UTF_8); /* fra KRR via SR */
    }

    @Override
    public String getOrgnrPostkasse() {
        return serviceRecord.getOrgnrPostkasse();
    }

    @Override
    public String getEmailAddress() {
        return serviceRecord.getEpostAdresse();
    }

    @Override
    public String getSmsVarslingstekst() {
        if (isDigitalMessage()) {
            return getDigitalMessage().getVarsler().getSmsTekst();
        }
        return props.getDpi().getEmail().getVarslingstekst();
    }

    @Override
    public String getEmailVarslingstekst() {
        if (isDigitalMessage()) {
            return getDigitalMessage().getVarsler().getEpostTekst();
        }
        return props.getDpi().getEmail().getVarslingstekst();
    }

    @Override
    public String getMobileNumber() {
        return serviceRecord.getMobilnummer();
    }

    @Override
    public boolean isNotifiable() {
        return serviceRecord.isKanVarsles();
    }

    @Override
    public boolean isPrintProvider() {
        return isPrintMessage();
    }

    @Override
    public PostAddress getPostAddress() {
        if (isPrintMessage()) {
            return getPrintMessage().getMottaker();
        }
        return null;
    }

    @Override
    public PostAddress getReturnAddress() {
        if (isPrintMessage()) {
            return getPrintMessage().getRetur().getMottaker();
        }
        return null;
    }

    @Override
    public Sikkerhetsnivaa getSecurityLevel() {
        if (isDigitalMessage()) {
            SDPSikkerhetsnivaa sdpSikkerhetsnivaa = SDPSikkerhetsnivaa.fromValue(String.valueOf(message.getBusinessMessage().getSikkerhetsnivaa()));
            return Sikkerhetsnivaa.valueOf(sdpSikkerhetsnivaa.toString());
        }
        return null;
    }

    @Override
    public Date getVirkningsdato() {
        if (isDigitalMessage()) {
            return Date.from(getDigitalMessage().getDigitalPostInfo()
                    .getVirkningsdato()
                    .atStartOfDay()
                    .atZone(clock.getZone())
                    .toInstant());
        }
        return new Date();
    }

    @Override
    public String getLanguage() {
        if (isDigitalMessage()) {
            return getDigitalMessage().getSpraak();
        }
        return props.getDpi().getLanguage();
    }

    @Override
    public boolean isAapningskvittering() {
        if (isDigitalMessage()) {
            return getDigitalMessage().getDigitalPostInfo().getAapningskvittering();
        }
        return false;
    }

    @Override
    public Utskriftsfarge getPrintColor() {
        if (isPrintMessage()) {
            return getPrintMessage().getUtskriftsfarge();
        }
        return Utskriftsfarge.SORT_HVIT;
    }

    @Override
    public Posttype getPosttype() {
        if (isPrintMessage()) {
            return getPrintMessage().getPosttype();
        }
        return Posttype.B_OEKONOMI;
    }

    @Override
    public Returhaandtering getReturnHandling() {
        if (isPrintMessage()) {
            return getPrintMessage().getRetur().getReturhaandtering();
        }
        return Returhaandtering.DIREKTE_RETUR;
    }
}

File: integrasjonspunkt/src/main/kotlin/no/difi/meldingsutveksling/nextmove/NextMoveDpiRequest.kt
package no.difi.meldingsutveksling.nextmove

import no.difi.begrep.sdp.schema_v10.SDPSikkerhetsnivaa
import no.difi.meldingsutveksling.api.OptionalCryptoMessagePersister
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties
import no.difi.meldingsutveksling.dpi.Document
import no.difi.meldingsutveksling.dpi.MeldingsformidlerRequest
import no.difi.meldingsutveksling.serviceregistry.externalmodel.ServiceRecord
import no.difi.sdp.client2.domain.MetadataDokument
import no.difi.sdp.client2.domain.digital_post.Sikkerhetsnivaa
import no.difi.sdp.client2.domain.fysisk_post.Posttype
import no.difi.sdp.client2.domain.fysisk_post.Returhaandtering
import no.difi.sdp.client2.domain.fysisk_post.Utskriftsfarge
import java.time.Clock
import java.util.*

class NextMoveDpiRequest(private val props: IntegrasjonspunktProperties,
                         private val clock: Clock,
                         private val message: NextMoveMessage,
                         private val serviceRecord: ServiceRecord,
                         private val optionalCryptoMessagePersister: OptionalCryptoMessagePersister) : MeldingsformidlerRequest {

    private fun createDocument(file: BusinessMessageFile): Document {
        val title = if (file.title.isNullOrBlank()) "Missing title" else file.title
        val document = Document(getContent(file.identifier), file.mimetype, file.filename, title)

        if (isDigitalMessage && getDigitalMessage().metadataFiler.containsKey(file.filename)) {
            val metadataFilename = getDigitalMessage().metadataFiler[file.filename]
            val metadataFile = message.files.first { it.filename == metadataFilename }
                    ?: throw NextMoveRuntimeException("Metadata document $metadataFilename specified for ${file.filename}, but is not attached")
            document.metadataDokument = MetadataDokument(metadataFilename, metadataFile.mimetype, getContent(metadataFile.identifier))
        }

        return document
    }

    private fun getContent(fileName: String): ByteArray {
        return optionalCryptoMessagePersister.read(message.messageId, fileName)
    }

    private fun isMetadataFile(filename: String): Boolean {
        return if (isPrintMessage) {
            false
        } else getDigitalMessage().metadataFiler.containsValue(filename)
    }

    private val isDigitalMessage: Boolean
        get() = message.businessMessage is DpiDigitalMessage

    private val isPrintMessage: Boolean
        get() = message.businessMessage is DpiPrintMessage

    private fun getDigitalMessage(): DpiDigitalMessage {
        return message.businessMessage as DpiDigitalMessage
    }

    private fun getPrintMessage(): DpiPrintMessage {
        return message.businessMessage as DpiPrintMessage
    }

    override fun getDocument(): Document {
        return message.files.first { it.primaryDocument }
                ?.let { createDocument(it) }
                ?: throw NextMoveRuntimeException("No primary documents found, aborting send")
    }

    override fun getAttachments(): List<Document> {
        return message.files.filter { !it.primaryDocument }
                .filter { !isMetadataFile(it.filename) }
                .map { createDocument(it) }
    }

    override fun getMottakerPid(): String {
        return message.receiverIdentifier
    }

    override fun getSubject(): String? {
        return when (val m = message.businessMessage) {
            is DpiDigitalMessage -> m.tittel
            is DpiPrintMessage -> null
            else -> throw NextMoveRuntimeException("BusinessMessage not instance of either ${DpiDigitalMessage::class.simpleName} or ${DpiPrintMessage::class.simpleName}")
        }
    }

    override fun getSenderOrgnumber(): String {
        return message.senderIdentifier
    }

    override fun getOnBehalfOfOrgnr(): Optional<String> {
        return message.sbd.partIdentifier
    }

    override fun getAvsenderIdentifikator(): Optional<String> {
        return when (val m = message.businessMessage) {
            is DpiDigitalMessage -> Optional.ofNullable(m.avsenderId)
            is DpiPrintMessage -> Optional.ofNullable(m.avsenderId)
            else -> Optional.empty()
        }
    }

    override fun getFakturaReferanse(): Optional<String> {
        return when (val m = message.businessMessage) {
            is DpiDigitalMessage -> Optional.ofNullable(m.fakturaReferanse)
            is DpiPrintMessage -> Optional.ofNullable(m.fakturaReferanse)
            else -> Optional.empty()
        }
    }

    override fun getConversationId(): String {
        return message.messageId
    }

    override fun getPostkasseAdresse(): String {
        return serviceRecord.postkasseAdresse
    }

    override fun getCertificate(): ByteArray {
        return serviceRecord.pemCertificate.toByteArray()
    }

    override fun getOrgnrPostkasse(): String {
        return serviceRecord.orgnrPostkasse
    }

    override fun getEmailAddress(): String? {
        return serviceRecord.epostAdresse
    }

    override fun getSmsVarslingstekst(): String? {
        return if (isDigitalMessage) getDigitalMessage().varsler?.smsTekst else null
    }

    override fun getEmailVarslingstekst(): String? {
        return if (isDigitalMessage) getDigitalMessage().varsler?.epostTekst else null
    }

    override fun getMobileNumber(): String? {
        return serviceRecord.mobilnummer
    }

    override fun isNotifiable(): Boolean {
        return serviceRecord.isKanVarsles
    }

    override fun isPrintProvider(): Boolean {
        return isPrintMessage
    }

    override fun getPostAddress(): PostAddress? {
        return if (isPrintMessage) {
            getPrintMessage().mottaker
        } else null
    }

    override fun getReturnAddress(): PostAddress? {
        return if (isPrintMessage) {
            getPrintMessage().retur.mottaker
        } else null
    }

    override fun getSecurityLevel(): Sikkerhetsnivaa? {
        return if (message.businessMessage is DpiDigitalMessage) {
            return Sikkerhetsnivaa.valueOf(SDPSikkerhetsnivaa.fromValue(message.businessMessage.getSikkerhetsnivaa().toString()).toString())
        } else null
    }

    override fun getVirkningsdato(): Date {
        return if (isDigitalMessage) {
            Date.from(getDigitalMessage().digitalPostInfo
                    .virkningsdato
                    .atStartOfDay()
                    .atZone(clock.zone)
                    .toInstant())
        } else Date()
    }

    override fun getLanguage(): String {
        return if (isDigitalMessage) {
            getDigitalMessage().spraak
        } else props.dpi.language
    }

    override fun isAapningskvittering(): Boolean {
        return if (isDigitalMessage) {
            getDigitalMessage().digitalPostInfo.aapningskvittering
        } else false
    }

    override fun getPrintColor(): Utskriftsfarge {
        return if (isPrintMessage) {
            getPrintMessage().utskriftsfarge
        } else Utskriftsfarge.SORT_HVIT
    }

    override fun getPosttype(): Posttype {
        return if (isPrintMessage) {
            getPrintMessage().posttype
        } else Posttype.B_OEKONOMI
    }

    override fun getReturnHandling(): Returhaandtering {
        return if (isPrintMessage) {
            getPrintMessage().retur.returhaandtering
        } else Returhaandtering.DIREKTE_RETUR
    }

}